File: local/migration/models/ir.py
"""Canonical intermediate representation for PAN-OS emission."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class Zone(BaseModel):
    name: str
    security_level: int | None = None


class AddressObject(BaseModel):
    name: str
    value: str  # ip-netmask CIDR or FQDN placeholder
    description: str | None = None


class AddressGroup(BaseModel):
    name: str
    members: list[str]
    static: bool = True


class ServiceObject(BaseModel):
    name: str
    protocol: str
    port: str | None = None  # single port, range "80-443", or None for protocol-only
    source_port: str | None = None


class ServiceGroup(BaseModel):
    name: str
    members: list[str]


class InterfaceConfig(BaseModel):
    asa_name: str
    pan_name: str
    zone: str | None = None
    ip_cidr: str | None = None
    mtu: int | None = None
    comment: str | None = None


class Route(BaseModel):
    virtual_router: str = "default"
    destination: str  # CIDR
    nexthop: str | None = None
    interface: str | None = None
    metric: int | None = None


class SecurityRule(BaseModel):
    name: str
    from_zones: list[str]
    to_zones: list[str]
    source: list[str]
    destination: list[str]
    service: list[str]
    application: list[str] = Field(default_factory=lambda: ["any"])
    action: Literal["allow", "deny", "drop"] = "allow"
    disabled: bool = False
    description: str | None = None
    log_start: bool = False
    log_end: bool = False


class NatRule(BaseModel):
    name: str
    from_zones: list[str]
    to_zones: list[str]
    source: list[str]
    destination: list[str]
    service: list[str] = Field(default_factory=lambda: ["any"])
    nat_type: Literal["static", "dynamic", "dynamicip", "identity"] = "dynamic"
    translated_source: str | None = None
    translated_destination: str | None = None
    translated_port: str | None = None
    bi_directional: bool = False


class VpnTunnel(BaseModel):
    name: str
    peer_ip: str
    ike_gateway_name: str
    ipsec_profile_name: str
    local_interface: str | None = None
    psk_placeholder: str = "[PSK_REMOVED]"
    transform_hint: str | None = None


class MigrationIR(BaseModel):
    hostname: str | None = None
    vsys: str = "vsys1"
    source_vendor: str = "cisco"
    zones: list[Zone] = Field(default_factory=list)
    addresses: list[AddressObject] = Field(default_factory=list)
    address_groups: list[AddressGroup] = Field(default_factory=list)
    services: list[ServiceObject] = Field(default_factory=list)
    service_groups: list[ServiceGroup] = Field(default_factory=list)
    interfaces: list[InterfaceConfig] = Field(default_factory=list)
    routes: list[Route] = Field(default_factory=list)
    security_rules: list[SecurityRule] = Field(default_factory=list)
    nat_rules: list[NatRule] = Field(default_factory=list)
    vpn_tunnels: list[VpnTunnel] = Field(default_factory=list)

File: dotnet/Services/Migration/Ir.cs
namespace PanCopilot.Services.Migration;

// Canonical intermediate representation for PAN-OS emission. Faithful port of
// migration/models/ir.py.

public sealed class Zone
{
    public string Name { get; set; } = "";
    public int? SecurityLevel { get; set; }
}

public sealed class AddressObject
{
    public string Name { get; set; } = "";
    public string Value { get; set; } = "";
    public string? Description { get; set; }
}

public sealed class AddressGroup
{
    public string Name { get; set; } = "";
    public List<string> Members { get; set; } = new();
    public bool Static { get; set; } = true;
}

public sealed class ServiceObject
{
    public string Name { get; set; } = "";
    public string Protocol { get; set; } = "tcp";
    public string? Port { get; set; }
    public string? SourcePort { get; set; }
}

public sealed class ServiceGroup
{
    public string Name { get; set; } = "";
    public List<string> Members { get; set; } = new();
}

public sealed class InterfaceConfig
{
    public string AsaName { get; set; } = "";
    public string PanName { get; set; } = "";
    public string? Zone { get; set; }
    public string? IpCidr { get; set; }
    public int? Mtu { get; set; }
    public string? Comment { get; set; }
}

public sealed class Route
{
    public string VirtualRouter { get; set; } = "default";
    public string Destination { get; set; } = "";
    public string? Nexthop { get; set; }
    public string? Interface { get; set; }
    public int? Metric { get; set; }
}

public sealed class SecurityRule
{
    public string Name { get; set; } = "";
    public List<string> FromZones { get; set; } = new();
    public List<string> ToZones { get; set; } = new();
    public List<string> Source { get; set; } = new();
    public List<string> Destination { get; set; } = new();
    public List<string> Service { get; set; } = new();
    public List<string> Application { get; set; } = new() { "any" };
    public string Action { get; set; } = "allow";   // allow | deny | drop
    public bool Disabled { get; set; }
    public string? Description { get; set; }
    public bool LogStart { get; set; }
    public bool LogEnd { get; set; }
}

public sealed class NatRule
{
    public string Name { get; set; } = "";
    public List<string> FromZones { get; set; } = new();
    public List<string> ToZones { get; set; } = new();
    public List<string> Source { get; set; } = new();
    public List<string> Destination { get; set; } = new();
    public List<string> Service { get; set; } = new() { "any" };
    public string NatType { get; set; } = "dynamic";  // static | dynamic | dynamicip | identity
    public string? TranslatedSource { get; set; }
    public string? TranslatedDestination { get; set; }
    public string? TranslatedPort { get; set; }
    public bool BiDirectional { get; set; }
}

public sealed class VpnTunnel
{
    public string Name { get; set; } = "";
    public string PeerIp { get; set; } = "";
    public string IkeGatewayName { get; set; } = "";
    public string IpsecProfileName { get; set; } = "";
    public string? LocalInterface { get; set; }
    public string PskPlaceholder { get; set; } = "[PSK_REMOVED]";
    public string? TransformHint { get; set; }
}

public sealed class MigrationIR
{
    public string? Hostname { get; set; }
    public string Vsys { get; set; } = "vsys1";
    public string SourceVendor { get; set; } = "cisco";
    public List<Zone> Zones { get; set; } = new();
    public List<AddressObject> Addresses { get; set; } = new();
    public List<AddressGroup> AddressGroups { get; set; } = new();
    public List<ServiceObject> Services { get; set; } = new();
    public List<ServiceGroup> ServiceGroups { get; set; } = new();
    public List<InterfaceConfig> Interfaces { get; set; } = new();
    public List<Route> Routes { get; set; } = new();
    public List<SecurityRule> SecurityRules { get; set; } = new();
    public List<NatRule> NatRules { get; set; } = new();
    public List<VpnTunnel> VpnTunnels { get; set; } = new();
}
